packages/explat-client/src/create-explat-client.ts

```diff
@@ -0,0 +1,150 @@
+	 */
+	loadExperimentAssignment: ( experimentName: string ) => Promise< ExperimentAssignment >;
+
+	/**
+	 * Get an already loaded Experiment Assignment, will throw if there is an error, e.g. if it hasn't been loaded.
+	 *
+	 * Make sure loadExperimentAssignment has been called before calling this function.
+	 *
+	 */
+	dangerouslyGetExperimentAssignment: ( experimentName: string ) => ExperimentAssignment;
+}
+
+export class MissingExperimentAssignmentError extends Error {
+	constructor( ...params: any[] ) {
+		super( ...params );
+
+		// Maintains proper stack trace for where our error was thrown (only available on V8)
+		if ( Error.captureStackTrace ) {
+			Error.captureStackTrace( this, MissingExperimentAssignmentError );
+		}
+
+		this.name = 'MissingExperimentAssignmentError';
+	}
+}
+
+/**
+ * Create an ExPlat Client
+ *
+ * @param config Configuration object
+ */
+export default function createExPlatClient( config: Config ): ExPlatClient {
+	/**
+	 * This bit of code is the heavy lifting behind loadExperimentAssignment, allowing it to be used intuitively.
+	 *
+	 * AOAAT stands for Async One At A Time, this is how we ensure for each experiment that there is only ever one fetch process occuring.
+	 *
+	 * @param experimentName The experiment's name
+	 */
+	const createAOAATExperimentAssignmentFetchAndStore = ( experimentName: string ) =>
+		Timing.asyncOneAtATime( async () => {
+			const fetchedExperimentAssignment = await Request.fetchExperimentAssignment(
+				config,
+				experimentName
+			);
+			State.storeExperimentAssignment( fetchedExperimentAssignment );
+			return fetchedExperimentAssignment;
+		} );
+	const experimentNameToAOAATExperimentAssignmentFetchAndStore: Record<
+		string,
+		() => Promise< ExperimentAssignment >
+	> = {};
+
+	return {
+		loadExperimentAssignment: async ( experimentName: string ): Promise< ExperimentAssignment > => {
+			try {
+				if ( ! Validation.isName( experimentName ) ) {
+					throw new Error( `Invalid experimentName: ${ experimentName }` );
+				}
+
+				const storedExperimentAssignment = State.retrieveExperimentAssignment( experimentName );
+				if (
+					storedExperimentAssignment &&
+					ExperimentAssignments.isAlive( storedExperimentAssignment )
+				) {
+					return storedExperimentAssignment;
+				}
+
+				if (
+					experimentNameToAOAATExperimentAssignmentFetchAndStore[ experimentName ] === undefined
+				) {
+					experimentNameToAOAATExperimentAssignmentFetchAndStore[
+						experimentName
+						// TODO: Move this outside of loadExperimentAssignment
+					] = createAOAATExperimentAssignmentFetchAndStore( experimentName );
+				}
+				// TODO: Move timeout within AOAAT
+				const fetchedExperimentAssignment = await Timing.timeoutPromise(
+					experimentNameToAOAATExperimentAssignmentFetchAndStore[ experimentName ](),
+					EXPERIMENT_FETCH_TIMEOUT
+				);
+				if ( ! fetchedExperimentAssignment ) {
+					throw new Error( `Could not fetch ExperimentAssignment` );
+				}
```

Comment:
Good point about offline mode! I hadn't properly considered it 😮 

Seems like returning the stale is a good idea in general.

Reply:
I have added it in with an additional check for the ExperimentAssignment to be at most 2 weeks old (I would prefer shorter, would that work?) so we keep an upper bound on them.

Reply:
> I have added it in with an additional check for the ExperimentAssignment to be at most 2 weeks old (I would prefer shorter, would that work?) so we keep an upper bound on them.

If we put a time limit on that, it would be the only client to do so. We shouldn't put a time limit on it, because doing so would be harmful to the user and solve nothing for us.

It's quite normal to start a post on an airplane, hang out at the GM, take a week off to explore the area, then open your laptop on the flight back. Things shouldn't stop working just because time has passed in real life while the laptop was suspended. It could introduce state corruption or any number of issues.

Reply:
> If we put a time limit on that, it would be the only client to do so. We shouldn't put a time limit on it, because doing so would be harmful to the user and solve nothing for us.

I've been thinking about this and I think it is an important safety mechanism:
- It prevents experiment experiences from running too long in these extreme offline cases (e.g. someone trapped without internet wanting to offline blog but stuck with a subpar treatment experience).
- It gives us an upper bound for how long experiments exist in the wild once we stop them - in certain situations it is important to know what is running and where.

> It's quite normal to start a post on an airplane, hang out at the GM, take a week off to explore the area, then open your laptop on the flight back. Things shouldn't stop working just because time has passed in real life while the laptop was suspended. 

The way I see this is that Experiments are always temporary, there will always be participants who experience a variation which then goes on to become discontinued. 

> It could introduce state corruption or any number of issues.

Experiments should also be designed to be stopped at any time - in the case they are disabled - they shouldn't cause any state corruption.

But... maybe recency check could be extended to three weeks or a month? My thinking is that two weeks without internet is a long time. Hmm, actually I need to look into how this client will work with the Desktop? Does it ever get reloaded?

Reply:
> It prevents experiment experiences from running too long in these extreme offline cases (e.g. someone trapped without internet wanting to offline blog but stuck with a bad treatment experience that we deliberately disabled early but just too late for that participant.).

This would happen with or without a deadline.

> It gives us an upper bound for how long experiments exist in the wild once we stop them - in certain situations it is important to know what is running and where.

What about for experiments that run for longer than two weeks?

It is extremely unlikely many people (if any) ever stay offline this long. However, we shouldn't remove features until they get online; regardless of whether the experiment is still running or not. It's ultimately not important (right now) to us that we know who is or isn't using an experiment offline, in fact, people may want to run experiments that run for offline users and we shouldn't break their experiment.

Reply:
Actually, I take that back.

> It is extremely unlikely many people (if any) ever stay offline this long.

I follow a few sailing blogs where they're offline for weeks/months at a time. It's extremely likely, but just a very small subset of users.

Reply:
Yeah these are good points, I'm going to remove it.